btf/types.go

```diff
@@ -67,13 +71,73 @@ func (ts types) ByID(id TypeID) (Type, error) {
 	return ts[id], nil
 }
 
+type Skipper func(Type) bool
+
+type Comparer func(Type, Type) bool
+
+// SkipQualifierAndTypedef skips qualifiers and Typedefs.
+var SkipQualifierAndTypedef = func(t Type) bool {
+	switch t.(type) {
+	case qualifier, *Typedef:
+		return true
+	}
+	return false
+}
+
+// CompareTypeName returns true if a and b's names correspond.
+var CompareTypeName = func(a, b Type) bool {
+	return a.TypeName() == b.TypeName()
+}
+
+// Equal compares Type a to Type b. By default, only binary equivalency is
+// checked, ignoring type names.
+//
+// cmp allows providing extra comparison rules. It will be evaluated after
+// checking a and b's binary equivalency at each step of the type hierarchy.
+// skip allows individual Types to be omitted from comparisons, and is called
+// before checking a and b's binary equivalency.
+func Equal(a, b Type, cmp Comparer, skip Skipper) error {
+	if a == nil && b == nil {
+		return nil
+	}
+	if a == nil || b == nil {
+		return errors.New("one of a or b are nil")
+	}
+
+	at, bt := postorderTraversal(a, nil), postorderTraversal(b, nil)
+	for at.Next() && bt.Next() {
+		if skip != nil {
+			for skip(at.Type) {
+				at.Next()
+			}
+			for skip(bt.Type) {
+				bt.Next()
+			}
+		}
+
+		if !at.Type.equal(bt.Type) {
+			return fmt.Errorf("a: %v and b: %v differ", at.Type, bt.Type)
+		}
+
+		if cmp != nil && !cmp(at.Type, bt.Type) {
+			return fmt.Errorf("a: %v and b: %v did not satisfy cmp", at.Type, bt.Type)
+		}
+	}
+
+	return nil
```

Comment:
This is missing a check for `!at.Empty() || !bt.Empty()`, see `coreAreTypesCompatible`.